<a href="https://colab.research.google.com/github/michalis0/DataMining_and_MachineLearning/blob/master/week6/Text_Analytics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Mining and Machine Learning - Week 6
# Text Analytics

[Text Analytics](https://people.ischool.berkeley.edu/~hearst/text-mining.html) (or text mining) is the process of deriving high-quality information from text. It involves "the discovery by computer of new, previously unknown information, by automatically extracting information from different written resources." Written resources may include websites, books, emails, reviews, and articles.

### Table of Contents
#### 1. Summary
* 1.1 Applications
* 1.2 Tokenization and Stopwords
* 1.3 Stemming and Lemmatization
* 1.4 Text Representation

#### 2. Text Preparation
* 2.1 Install spaCy
* 2.2 Tokenization
* 2.3 Dependency Parsing
* 2.4 Remove Stopwords
* 2.5 Lemmatization
* 2.6 Entity Detection 

#### 3. Text Representation
* 3.1 Bag of Words (BOW)
* 3.2 TF-IDF Representation

#### 4. Text Classification: Alexa Reviews
* 4.1 Load and prepare data
* 4.2 Classification of the reviews using logistic regression
* 4.3 BONUS

## 1. Summary

### 1.1 Applications
There are many applications of text analytics, for example:
* Search for relevant websites or articles using a search engine
* Sentiment Analysis (e.g. classify tweets or film reviews as positive, neutral or negative)
* Chatbots (e.g. Siri, Alexa)
* Project idea: The Impact of Donald Trump’s Tweets on Financial Market
* Etc.

### 1.2 Tokenization and Stopwords
Tokens are the elementary building blocks (words, numbers, characters) in a document. Tokenization is the process of splitting an input
sequence into tokens. Example: "I love data science" --> "I", "love", "data", "science". Stopwords are common words that appear very frequently (e.g. "is", "and", "you", etc.). It is convenient to remove them as they do not add much to the content of a document and are therefore generally not useful for text analysis or, worse still, make it worse.

### 1.3 Lemmatization and Stemming
* Goal: have the same token for different forms of a word (e.g. fishing, fished, fisher, fishers, etc.)
* Lemmatization: Find what is the lemma of a word (e.g. feet -> foot)
* Stemming: one method for lemmatization where rules that remove the ending of a word are applied (e.g. fishing -> fish)


### 1.4 Text Representation
* Goal: transform text such that it can be used for text analysis
* Bag of Words (BOW): works in many case but order is not preserved (solution: n-grams)
* TF-IDF: emphasizes important words

## 2. Text Preparation
In this section, we explain how to prepare a text for analysis. This includes tockeninzing the text, removing stopwords, etc.

### 2.1 Install spaCy
[spaCy](https://spacy.io/) is an open-source natural language processing library for Python. It is designed particularly for production use, and it can help us to build applications that process massive volumes of text efficiently.

We install the library and its English-language model.

In [1]:
# Install and update spaCy
!pip install -U spacy

# Load the english language model
!python -m spacy download en

Requirement already up-to-date: spacy in /usr/local/lib/python3.6/dist-packages (2.3.2)
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_sm')
✔ Linking successful
/usr/local/lib/python3.6/dist-packages/en_core_web_sm -->
/usr/local/lib/python3.6/dist-packages/spacy/data/en
You can now load the model via spacy.load('en')


In [2]:
# Import required packages
import spacy
from spacy import displacy
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import pandas as pd

### 2.2 Tokenization

Tokenization is the process of breaking a text into pieces called tokens. A token simply refers to an individual part of a sentence having some semantic value. SpaCy‘s tokenizer takes input in form of unicode text and outputs a sequence of token objects. In addition, SpaCy automatically breaks your document into tokens when a document is created using the language model.

Let’s take a look at a simple example. Imagine we have the following text, and we would like to tokenize it:

> When learning data science, you shouldn't get discouraged!

> Challenges and setbacks aren't failures, they're just part of the journey. You've got this!

There are a couple of different ways we can appoach this. The first is called __word tokenization__, which means breaking up the text into individual words. This is a critical step for many language processing applications, as they often require inputs in the form of individual words rather than longer strings of text.

In [3]:
# Load English language model
sp = spacy.load('en_core_web_sm')

# Declare the text
text = """When learning data science, you shouldn't get discouraged!
Challenges and setbacks aren't failures, they're just part of the journey. You've got this!"""

# spaCy object is used to create a document
my_doc = sp(text)

my_doc

When learning data science, you shouldn't get discouraged!
Challenges and setbacks aren't failures, they're just part of the journey. You've got this!

In [4]:
type(my_doc)

spacy.tokens.doc.Doc

In [5]:
# Create list of word tokens
token_list = []

for token in my_doc:
    token_list.append(token.text)

token_list

['When',
 'learning',
 'data',
 'science',
 ',',
 'you',
 'should',
 "n't",
 'get',
 'discouraged',
 '!',
 '\n',
 'Challenges',
 'and',
 'setbacks',
 'are',
 "n't",
 'failures',
 ',',
 'they',
 "'re",
 'just',
 'part',
 'of',
 'the',
 'journey',
 '.',
 'You',
 "'ve",
 'got',
 'this',
 '!']


As we can see, spaCy produces a list that contains each token as a separate item. Notice that it has recognized that contractions such as _shouldn’t_ actually represent two distinct words, and has thus broken them down into two distinct tokens.

In the example above, we first load the language dictionary. Here we load the english one using `spacy.load('en_core_web_sm')` create an object of this class, "sp", which is used to create documents with linguistic annotations and various language properties. After creating the document, we create a list of tokens.

We can also see the parts-of-speech (POS) of each of these tokens using the `.pos_` attribute, as shown below. POS tagging can be really useful, particularly if you have words or tokens that can have multiple POS tags. For instance, the word "fish" can be used as both a noun and verb, depending upon the context.

In [6]:
# POS
for word in my_doc:
    print(word.text, word.pos_)

When ADV
learning VERB
data NOUN
science NOUN
, PUNCT
you PRON
should VERB
n't PART
get AUX
discouraged ADJ
! PUNCT

 SPACE
Challenges NOUN
and CCONJ
setbacks NOUN
are AUX
n't PART
failures NOUN
, PUNCT
they PRON
're AUX
just ADV
part NOUN
of ADP
the DET
journey NOUN
. PUNCT
You PRON
've AUX
got VERB
this DET
! PUNCT


In [7]:
# Another example
doc1 = sp("I like to fish") # verb
doc2 = sp("I eat a fish") # noun

for word in doc1:
  print(word.text, word.pos_)

print("-----------------")

for word in doc2:
  print(word.text, word.pos_)

I PRON
like VERB
to PART
fish VERB
-----------------
I PRON
eat VERB
a DET
fish NOUN



If we want, we can also break the text into sentences rather than words. This is called __sentence tokenization__. When performing sentence tokenization, the tokenizer looks for specific characters that normally fall between sentences, like periods, exclaimation points, and newline characters.

In [8]:
# create list of sentence tokens
sents_list = []

for sent in my_doc.sents:
    sents_list.append(sent.text)

sents_list

["When learning data science, you shouldn't get discouraged!\n",
 "Challenges and setbacks aren't failures, they're just part of the journey.",
 "You've got this!"]

### 2.3 Dependency Parsing
__Depenency parsing__ is a language processing technique that allows to better determine the meaning of a sentence by analyzing how it is constructed to determine how the individual words relate to each other.

Consider, for example, the sentence “Joe throws the ball.” We have two nouns (Joe and ball) and one verb (throws). But we can’t just look at these words individually, or we may end up thinking that the ball throws Joe! To understand the sentence correctly, we need to look at the word order and sentence structure, not just the words.

Below, we have a short sentence. We’ll use a spaCy method called `noun_chunks`, which breaks the input down into nouns and the words describing them, and iterate through each chunk in our source text, identifying the word, its root, its dependency identification, and which chunk it belongs to.

In [9]:
doc = sp(" Joe threw a ball, and President Donald, in pursuit of the ball, hit a wall.") # notice the space at the beginning

for chunk in doc.noun_chunks:
    print(chunk.text, chunk.root.text, chunk.root.dep_,
          chunk.root.head.text)

 Joe Joe nsubj threw
a ball ball dobj threw
President Donald Donald nsubj hit
pursuit pursuit pobj in
the ball ball pobj of
a wall wall dobj hit


In [10]:
# Let's visualize this
displacy.render(doc, style="dep", jupyter= True, options={'distance': 120})

### 2.4 Remove Stopwords
Most text data that we work with is going to contain a lot of words that are not actually useful to us. These words, called stopwords, are useful in human speech, but they do not have much to contribute to data analysis. Removing stopwords helps us eliminate noise and distraction from our text data, and also speeds up the time analysis takes (since there are fewer words to process). This makes text analysis more efficient.

Let’s take a look at the stopwords spaCy includes by default.

In [11]:
# Import stopwords from English language
spacy_stopwords = spacy.lang.en.stop_words.STOP_WORDS

# Print total number of stopwords
print('Number of stopwords: %d' % len(spacy_stopwords))

# Print 20 stopwords
print('20 stopwords: %s' % list(spacy_stopwords)[:20])

Number of stopwords: 326
20 stopwords: ['anywhere', 'above', 'both', 'last', 'otherwise', 'thence', 'does', 'seems', 'with', 'so', 'indeed', 'herself', 'a', 'nobody', 'whereby', 'whither', 'were', 'call', 'except', 'anyone']


Now that we’ve got our list of stopwords, let’s use it to remove the stopwords from the text string we were working on in the previous section.

In [12]:
# Wgich words will be removed?
my_doc

When learning data science, you shouldn't get discouraged!
Challenges and setbacks aren't failures, they're just part of the journey. You've got this!

In [13]:
# Declare list for filtered sentence
filtered_sent = []

# Filter stopwords
for word in my_doc:
    if word.is_stop == False:
        filtered_sent.append(word)

filtered_sent

[learning,
 data,
 science,
 ,,
 discouraged,
 !,
 ,
 Challenges,
 setbacks,
 failures,
 ,,
 journey,
 .,
 got,
 !]

In [14]:
# We can also remove the punctuation
filtered_sent2 = []
removed_tokens = []

# Filter stopwords, punctuation and spaces
for word in my_doc:
  if (word.is_stop == True) or (word.is_punct == True) or (word.is_space == True):
    removed_tokens.append(word)
  else:
    filtered_sent2.append(word)

removed_tokens

[When,
 ,,
 you,
 should,
 n't,
 get,
 !,
 ,
 and,
 are,
 n't,
 ,,
 they,
 're,
 just,
 part,
 of,
 the,
 .,
 You,
 've,
 this,
 !]

In [15]:
filtered_sent2

[learning,
 data,
 science,
 discouraged,
 Challenges,
 setbacks,
 failures,
 journey,
 got]

### 2.5 Lemmatization
Lemmatization is a way of dealing with the fact that while words like connect, connection, connecting, connected, etc. aren’t exactly the same, they all have the same essential meaning: connect. The differences in spelling have grammatical functions in spoken language, but for machine processing, those differences can be confusing, so we need a way to change all the words that are forms of the word connect into the word connect itself.

One method for doing this is called __stemming__. Stemming involves simply lopping off easily-identified prefixes and suffixes to produce what’s often the simplest version of a word, the root. Connection, for example, would have the -ion suffix removed and be correctly reduced to connect. This kind of simple stemming is often all that’s needed, but lemmatization—which actually looks at words and their roots (called lemma) as described in the dictionary—is more precise (e.g feet -> foot).

Let's look at this simple example.

In [16]:
# Lemmatization
lem = sp("run runs ran running runner runners")

# Find lemma for each word
for word in lem:
    print(word.text, word.lemma_)

run run
runs run
ran run
running run
runner runner
runners runner


### 2.6 Entity Detection

__Entity detection__, also called entity recognition, is a more advanced form of language processing that identifies important elements like places, people, organizations, and languages within a text. This is really helpful for quickly extracting information from the text, since you can quickly pick out important topics or indentify key sections of it.

Let’s try out some entity detection using a few paragraphs from this [article](https://www.bloomberg.com/features/trump-tweets-market/).

In [17]:
article = sp("""
President Donald Trump gets a lot of attention for using Twitter to attack American trading partners, political foes, and media companies. But he often takes to the platform to celebrate the strength of the world’s largest economy and its publicly-traded companies.

Before U.S. stocks peaked in late January, he drew a direct connection between the increase in market value of American companies and his administration’s pro-growth policies on more than 10 occasions in that month alone.
""")

entities = [(i, i.label_, i.label) for i in article.ents]
entities

[(Donald Trump, 'PERSON', 380),
 (Twitter, 'ORG', 383),
 (American, 'NORP', 381),
 (U.S., 'GPE', 384),
 (late January, 'DATE', 391),
 (American, 'NORP', 381),
 (more than 10, 'CARDINAL', 397),
 (that month alone, 'DATE', 391)]

The above example how spaCy is able to identify a variety of different entity types, including specific locations (GPE), date-related words (DATE), important numbers (CARDINAL), specific individuals (PERSON), etc.

Using `displaCy` we can also visualize the text, with each identified entity highlighted by a color and labeled. We’ll use `style = "ent"` to tell displaCy that we want to visualize entities here.

In [18]:
displacy.render(article, style = "ent", jupyter = True)

## 3. Text Representation
We now show how to transform a text into an usable input for text classification. We use the first sentence of the article from the last section and two other sentences.

In [19]:
# Article as a string, not a spacy object
article = """
President Donald Trump gets a lot of attention for using Twitter to attack American trading partners, political foes, and media companies."""

# Sentences
s1 = """Donald Trump is a great friend, and he has four or five Picassos on his plane. And that's where I would look at them.""" # from Shaquille O'Neal
s2 = """Donald Trump is a phony, a fraud. His promises are as worthless as a degree from Trump University.""" # from Mitt Romney

# List of sentences
texts = [article, s1, s2]

### 3.1 Bag of Words (BOW)
We use the [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) class of scikit learn.

In [20]:
# Using default tokenizer 
count = CountVectorizer(ngram_range=(1,2), stop_words="english")
bow = count.fit_transform(texts)

# Show feature matrix
bow.toarray()

array([[1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
        1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1,
        0, 0, 0, 1, 1, 0, 1, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
        1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 2, 0,
        0, 1, 1, 0, 0, 1, 0, 0, 1, 1]])

In [21]:
# Get feature names
feature_names = count.get_feature_names()

# View feature names
feature_names

['american',
 'american trading',
 'attack',
 'attack american',
 'attention',
 'attention using',
 'companies',
 'degree',
 'degree trump',
 'donald',
 'donald trump',
 'foes',
 'foes media',
 'fraud',
 'fraud promises',
 'friend',
 'friend picassos',
 'gets',
 'gets lot',
 'great',
 'great friend',
 'look',
 'lot',
 'lot attention',
 'media',
 'media companies',
 'partners',
 'partners political',
 'phony',
 'phony fraud',
 'picassos',
 'picassos plane',
 'plane',
 'plane look',
 'political',
 'political foes',
 'president',
 'president donald',
 'promises',
 'promises worthless',
 'trading',
 'trading partners',
 'trump',
 'trump gets',
 'trump great',
 'trump phony',
 'trump university',
 'twitter',
 'twitter attack',
 'university',
 'using',
 'using twitter',
 'worthless',
 'worthless degree']

In [22]:
# Show as a dataframe
pd.DataFrame(
    bow.todense(), 
    columns=feature_names
    )

american  american trading  ...  worthless  worthless degree
0         1                 1  ...          0                 0
1         0                 0  ...          0                 0
2         0                 0  ...          1                 1

[3 rows x 54 columns]

### 3.2 TF-IDF Representation


Recall that:

- term frequency tf = count(word, document) / len(document) 
- term frequency idf = log( len(collection) / count(document_containing_term, collection) )
- tf-idf = tf * idf 

It is important to mention that the IDF value for a word remains the same throughout all the documents as it depends upon the total number of documents. On the other hand, TF values of a word differ from document to document.

**Note**: In the example below, you may not get the exact values by multiplying those two numbers, because nltk normalizes each row to have norm of 1. However the relative importance of the terms won't change.


In [23]:
# Using default tokenizer in TfidfVectorizer
tfidf = TfidfVectorizer(ngram_range=(1, 1), stop_words="english")
features = tfidf.fit_transform(texts)
pd.DataFrame(
    features.todense(),
    columns=tfidf.get_feature_names()
)

american    attack  attention  ...  university     using  worthless
0  0.260841  0.260841   0.260841  ...    0.000000  0.260841   0.000000
1  0.000000  0.000000   0.000000  ...    0.000000  0.000000   0.000000
2  0.000000  0.000000   0.000000  ...    0.359347  0.000000   0.359347

[3 rows x 27 columns]

In [24]:
texts

['\nPresident Donald Trump gets a lot of attention for using Twitter to attack American trading partners, political foes, and media companies.',
 "Donald Trump is a great friend, and he has four or five Picassos on his plane. And that's where I would look at them.",
 'Donald Trump is a phony, a fraud. His promises are as worthless as a degree from Trump University.']

## 4. Text Classification: Alexa reviews

We are going to use a real-world data set: [Amazon Alexa product reviews](https://www.kaggle.com/sid321axn/amazon-alexa-reviews/download).

This data set comes as a tab-separated file (.tsv). It has has five columns: `rating`, `date`, `variation`, `verified_reviews`, `feedback`.

`rating` denotes the rating each user gave Alexa (out of 5). `date` indicates the date of the review, and `variation` describes which model the user reviewed. `verified_reviews` contains the text of each review, and `feedback` contains a sentiment label, with 1 denoting positive sentiment (the user liked it) and 0 denoting negative sentiment (the user didn’t).

We are going develop a classification model that looks at the review text and predicts whether a review is positive or negative. Since this data set already includes whether a review is positive or negative in the `feedback` column, we can use those answers to train and test our model. Our goal here is to produce an accurate model that we could then use to process new user reviews and quickly determine whether they were positive or negative.

### 4.1 Load and prepare data

In [25]:
# Import additional packages
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline
import string
from spacy.lang.en.stop_words import STOP_WORDS
from spacy.lang.en import English
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

In [26]:
# Load data
url = "https://raw.githubusercontent.com/michalis0/DataMining_and_MachineLearning/master/week6/data/amazon_alexa.tsv"
df = pd.read_csv(url, delimiter="\t")
df.sample(10)

rating  ... feedback
1401       5  ...        1
1789       5  ...        1
2163       5  ...        1
2306       5  ...        1
2097       5  ...        1
1569       4  ...        1
2737       5  ...        1
2109       5  ...        1
2797       5  ...        1
654        5  ...        1

[10 rows x 5 columns]

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   rating            3150 non-null   int64 
 1   date              3150 non-null   object
 2   variation         3150 non-null   object
 3   verified_reviews  3150 non-null   object
 4   feedback          3150 non-null   int64 
dtypes: int64(2), object(3)
memory usage: 123.2+ KB


In [28]:
# Change date to datetime
df["date"] = pd.to_datetime(df["date"])

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   rating            3150 non-null   int64         
 1   date              3150 non-null   datetime64[ns]
 2   variation         3150 non-null   object        
 3   verified_reviews  3150 non-null   object        
 4   feedback          3150 non-null   int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 123.2+ KB


In [30]:
# Base rate: the data-set is unbalanced!
df.feedback.value_counts()

1    2893
0     257
Name: feedback, dtype: int64

In [31]:
round(df.feedback.value_counts()[1] / len(df), 4)

0.9184

###### Tokening the Data With spaCy

We create a `spacy_tokenizer()` function that accepts a sentence as input and processes the sentence into tokens, performing lemmatization, lowercasing, and removing stopwords.

__A note from spacy documentation__: spaCy adds a special case for pronouns: all pronouns are lemmatized to the special token `-PRON-`. Unlike verbs and common nouns, there’s no clear base form of a personal pronoun. Should the lemma of “me” be “I”, or should we normalize person as well, giving “it” — or maybe “he”? spaCy’s solution is to introduce a novel symbol, `-PRON-`, which is used as the lemma for all personal pronouns.

In [32]:
# Create a list of punctuation marks
punctuations = string.punctuation

punctuations

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [33]:
# Create a list of stopwords
stop_words = spacy.lang.en.stop_words.STOP_WORDS

list(stop_words)[:10]

['anywhere',
 'above',
 'both',
 'last',
 'otherwise',
 'thence',
 'does',
 'seems',
 'with',
 'so']

In [34]:
# Load English language model
sp = spacy.load('en_core_web_sm')

# Create tokenizer function
def spacy_tokenizer(sentence):
    # Create token object, which is used to create documents with linguistic annotations.
    mytokens = sp(sentence)

    # Lemmatize each token and convert each token into lowercase
    mytokens = [ word.lemma_.lower().strip() if word.lemma_ != "-PRON-" else word.lower_ for word in mytokens ]

    # Remove stop words and punctuation
    mytokens = [ word for word in mytokens if word not in stop_words and word not in punctuations ]

    # Return preprocessed list of tokens
    return mytokens

# Example
review = df["verified_reviews"].sample()
review.values[0]

'The product sounded the same as the emoji speaker from five below my sister has ... and even that one has Bluetooth and doesn’t need to be plugged in. The only good thing about this is that you can speak to it.'

In [35]:
spacy_tokenizer(review.values[0])

['product',
 'sound',
 'emoji',
 'speaker',
 'sister',
 '...',
 'bluetooth',
 'need',
 'plug',
 'good',
 'thing',
 'speak']

#### Vectorization Feature Engineering (TF-IDF)

We use the TF-IDF (Term Frequency-Inverse Document Frequency) to vectorize the documents. This is a way of representing how important a particular term is in the context of a given document, based on how many times the term appears and how many other documents that same term appears in. The higher the TF-IDF, the more important that term is to that document.

In [36]:
tfidf_vector = TfidfVectorizer(tokenizer=spacy_tokenizer) # we use the above defined tokenizer

### 4.2 Classification of the reviews using logistic regression

In [37]:
# Select features
X = df['verified_reviews'] # the features we want to analyze
ylabels = df['feedback'] # the labels, or answers, we want to test against

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, ylabels, test_size=0.2, random_state=1234)

X_train

2088    I really like this item. I already had the ech...
287     Technology in such a small device. The price w...
516           Pleased with the order just what I wanted!!
832                                               Love it
890                                   Works as advertised
                              ...                        
664     Great for mom its inside an owlhead in her jac...
3125    This product is easy to use and very entertain...
1318    You're going to have to shell out an extra $20...
723     Fun item to play with and get used to using.  ...
2863                                        Great product
Name: verified_reviews, Length: 2520, dtype: object

In [38]:
y_train

2088    1
287     1
516     1
832     1
890     1
       ..
664     1
3125    1
1318    1
723     1
2863    1
Name: feedback, Length: 2520, dtype: int64

In [39]:
# Define classifier
classifier = LogisticRegression()

# Create pipeline
pipe = Pipeline([('vectorizer', tfidf_vector),
                 ('classifier', classifier)])

# Fit model on training set
pipe.fit(X_train, y_train)

Pipeline(memory=None,
         steps=[('vectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_patt...,
                                 tokenizer=<function spacy_tokenizer at 0x7fedd1b6cd90>,
                                 use_idf=True, vocabulary=None)),
                ('classifier',
                 LogisticRegression(C=1.0, class_weight=None, dual=False,
                                    fit_i

In [40]:
# Evaluate the model
def evaluate(true, pred):
    precision = precision_score(true, pred)
    recall = recall_score(true, pred)
    f1 = f1_score(true, pred)
    print(f"CONFUSION MATRIX:\n{confusion_matrix(true, pred)}")
    print(f"ACCURACY SCORE:\n{accuracy_score(true, pred):.4f}")
    print(f"CLASSIFICATION REPORT:\n\tPrecision: {precision:.4f}\n\tRecall: {recall:.4f}\n\tF1_Score: {f1:.4f}")

In [41]:
# Predictions
y_pred = pipe.predict(X_test)

# Evaluation - test set
evaluate(y_test, y_pred)

CONFUSION MATRIX:
[[  2  49]
 [  0 579]]
ACCURACY SCORE:
0.9222
CLASSIFICATION REPORT:
	Precision: 0.9220
	Recall: 1.0000
	F1_Score: 0.9594


For the test set, the model correctly identifies a sentiment 92.22% of the time. This is only slightly better than the base rate (91.84%). The difference lies in the two cases that the model correctly classifies as negative. Therefore, the model does not work very well. Maybe because it has too many positive reviews and cannot identify negative one among them.

A recall of 1 means that if a sentiment is positive, it will be classififed as positive.

We observe approximately the same on the training set, as shown below.

In [42]:
# Evaluate on training set
evaluate(y_train, pipe.predict(X_train))

CONFUSION MATRIX:
[[   6  200]
 [   0 2314]]
ACCURACY SCORE:
0.9206
CLASSIFICATION REPORT:
	Precision: 0.9204
	Recall: 1.0000
	F1_Score: 0.9586


In [43]:
# Which reviews are classified as negative in test set?
X_test.loc[pipe.predict(X_test) == 0]

491                            I've already returned it.
381    It worked for a month or so then it stopped. I...
Name: verified_reviews, dtype: object

In [44]:
# Which reviews are classified as negaive in training set?
X_train.loc[pipe.predict(X_train) == 0]

1348    Can't turn of &#34;things to try&#34; on the s...
2611    This worked well for about 6 months but then s...
551     product stopped working after return time ran out
420     Item has never worked. Out of box it is broken...
1903             Terrible. Stopped working after one day.
2962    This worked well for about 6 months but then s...
Name: verified_reviews, dtype: object

In [45]:
# Prediction for new reviews
example_review_1 = "I really love the product. It is very helpful. I use it everyday" # positive
example_review_2 = "It stopped working, I want to return it" # negative
example_review_3 = "I don't like it, it is bad" # negative

examples = pd.Series([example_review_1, example_review_2, example_review_3])
examples

0    I really love the product. It is very helpful....
1              It stopped working, I want to return it
2                           I don't like it, it is bad
dtype: object

In [46]:
pipe.predict(examples)

array([1, 0, 1])

### 4.3 BONUS
In order to improve the prediction, we can:
* Tun the hyperparameters
* Use another classififer
* Improve text preparation

We illustrate the use of another classifier below. In addition, we try to predict the rating.

In [47]:
# BONUS 1: how to improve the classification: 
# --> tun your classififer or use another one
from sklearn.ensemble import RandomForestClassifier

# Define classifier
classifier = RandomForestClassifier(n_estimators=50)

# Create pipeline
pipe = Pipeline([('vectorizer', tfidf_vector),
                 ('classifier', classifier)])

# Fit model on training set
pipe.fit(X_train, y_train)

# Predictions
y_pred = pipe.predict(X_test)

# Evaluation - test set
evaluate(y_test, y_pred)
print("-----------------------")

# Evaluation - training set
evaluate(y_train, pipe.predict(X_train))

CONFUSION MATRIX:
[[  9  42]
 [  0 579]]
ACCURACY SCORE:
0.9333
CLASSIFICATION REPORT:
	Precision: 0.9324
	Recall: 1.0000
	F1_Score: 0.9650
-----------------------
CONFUSION MATRIX:
[[ 189   17]
 [   0 2314]]
ACCURACY SCORE:
0.9933
CLASSIFICATION REPORT:
	Precision: 0.9927
	Recall: 1.0000
	F1_Score: 0.9963


In [48]:
# Which reviews are classified as negative in test set?
X_test.loc[pipe.predict(X_test) == 0].values

array(["I am not super impressed with Alexa. When my Prime lapsed, she wouldn't play anything. She isn't smart enough to differentiate among spotify accounts so we can't use it for that either. She randomly speaks up when nobody is talking to her. Just today I unplugged her...not sure I'll ever use my Alexa again.",
       "It's got great sound and bass but it doesn't work all of the time. Its still hot or miss when it recognizes things",
       'The only negative we have on this product is the terrible sound quality.  A massive difference from the Alexa.  Which to us was a big reason we wanted to purchase this.Won’t be buying another until the speaker and sound quality can improve.',
       "I've already returned it.",
       'Returned from repair with No repair done. It has a problem which requires it to be on for a while for the defective part to show itself.I sent it for repair and got it back without being fixed.  The site gives no option to complain from the page.I have a echo sp

In [49]:
# BONUS 2: predict the rating
df.sample(5)

rating  ... feedback
44         5  ...        1
1180       5  ...        1
2925       4  ...        1
2784       5  ...        1
1495       5  ...        1

[5 rows x 5 columns]

In [50]:
df.rating.value_counts()

5    2286
4     455
1     161
3     152
2      96
Name: rating, dtype: int64

In [51]:
# Base rate
round(df.rating.value_counts()[5] / len(df), 4)

0.7257

In [52]:
# Select features
X = df['verified_reviews'] # the features we want to analyze
y = df['rating'] # the labels, or answers, we want to test against

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1234)

# Define classifier
classifier = LogisticRegression()

# Create pipeline
pipe = Pipeline([('vectorizer', tfidf_vector),
                 ('classifier', classifier)])

# Generate Model on training set
pipe.fit(X_train, y_train)

# Predictions
y_pred = pipe.predict(X_test)

# Evaluation - test set
print(f"CONFUSION MATRIX:\n{confusion_matrix(y_test, y_pred)}")
print(f"ACCURACY SCORE:\n{accuracy_score(y_test, y_pred):.4f}")

CONFUSION MATRIX:
[[  2   0   0   4  31]
 [  0   1   0   2  11]
 [  0   0   0   3  32]
 [  0   0   0   5  87]
 [  0   0   0   5 447]]
ACCURACY SCORE:
0.7222


In [53]:
# BONUS 3: use random forest

# Define classifier
classifier = RandomForestClassifier(n_estimators=50)

# Create pipeline
pipe = Pipeline([('vectorizer', tfidf_vector),
                 ('classifier', classifier)])

# Generate Model on training set
pipe.fit(X_train, y_train)

# Predictions
y_pred = pipe.predict(X_test)

# Evaluation - test set
print(f"CONFUSION MATRIX:\n{confusion_matrix(y_test, y_pred)}")
print(f"ACCURACY SCORE:\n{accuracy_score(y_test, y_pred):.4f}")

CONFUSION MATRIX:
[[  7   1   1   2  26]
 [  0   4   0   0  10]
 [  0   0  12   4  19]
 [  0   0   0  34  58]
 [  0   0   1   3 448]]
ACCURACY SCORE:
0.8016
